src/modules/job-manager/alloc.c

```diff
@@ -232,6 +234,43 @@ int cancel_request (struct alloc *alloc, struct job *job)
     return 0;
 }
 
+static void update_annotations (flux_t *h, struct job *job, flux_jobid_t id,
+                                json_t *annotations)
+{
+    if (annotations) {
+        if (!job->annotations) {
+            if (!(job->annotations = json_object ()))
+                flux_log (h,
+                          LOG_ERR,
+                          "%s: id=%ju json_object",
+                          __FUNCTION__, (uintmax_t)id);
+        }
+        if (job->annotations) {
+            const char *key;
+            json_t *value;
+
+            json_object_foreach (annotations, key, value) {
+                if (!json_is_null (value)) {
+                    if (json_object_set (job->annotations, key, value) < 0)
+                        flux_log (h,
+                                  LOG_ERR,
+                                  "%s: id=%ju update key=%s",
+                                  __FUNCTION__, (uintmax_t)id, key);
+                }
+                else
+                    /* not an error if key doesn't exist in job->annotations */
+                    (void)json_object_del (job->annotations, key);
+            }
+        }
+        /* Special case: if user cleared all entries, assume we no
```

Comment:
i just realized, this should go inside the branch statement above this.  Got a fixup for it already.